indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceInputFormat.java

```diff
@@ -191,64 +187,58 @@ private DatasourceInputSplit toDataSourceSplit(
       JobConf conf
   )
   {
-    String[] locations = null;
-    try {
-      locations = getFrequentLocations(segments, fio, conf);
-    }
-    catch (Exception e) {
-      logger.error(e, "Exception thrown finding location of splits");
-    }
+    String[] locations = getFrequentLocations(segments, fio, conf);
+
     return new DatasourceInputSplit(segments, locations);
   }
 
   private String[] getFrequentLocations(
-      List<WindowedDataSegment> segments,
-      org.apache.hadoop.mapred.InputFormat fio,
-      JobConf conf
-  ) throws IOException
-  {
-    Iterable<String> locations = Collections.emptyList();
-    for (WindowedDataSegment segment : segments) {
-      FileInputFormat.setInputPaths(conf, new Path(JobHelper.getURIFromSegment(segment.getSegment())));
-      for (org.apache.hadoop.mapred.InputSplit split : fio.getSplits(conf, 1)) {
-        locations = Iterables.concat(locations, Arrays.asList(split.getLocations()));
-      }
-    }
-    return getFrequentLocations(locations);
-  }
-
-  private static String[] getFrequentLocations(Iterable<String> hosts)
+      final List<WindowedDataSegment> segments,
+      final org.apache.hadoop.mapred.InputFormat fio,
+      final JobConf conf
+  )
   {
-
-    final CountingMap<String> counter = new CountingMap<>();
-    for (String location : hosts) {
-      counter.add(location, 1);
-    }
-
-    final TreeSet<Pair<Long, String>> sorted = Sets.<Pair<Long, String>>newTreeSet(
-        new Comparator<Pair<Long, String>>()
-        {
-          @Override
-          public int compare(Pair<Long, String> o1, Pair<Long, String> o2)
-          {
-            int compare = o2.lhs.compareTo(o1.lhs); // descending
-            if (compare == 0) {
-              compare = o1.rhs.compareTo(o2.rhs);   // ascending
-            }
-            return compare;
+    final Stream<String> locations = segments.stream().flatMap(
+        (final WindowedDataSegment segment) -> {
+          FileInputFormat.setInputPaths(
+              conf,
+              new Path(JobHelper.getURIFromSegment(segment.getSegment()))
+          );
+          try {
+            return Arrays.stream(fio.getSplits(conf, 1)).flatMap(
+                (final org.apache.hadoop.mapred.InputSplit split) -> {
+                  try {
+                    return Arrays.stream(split.getLocations());
+                  }
+                  catch (final IOException e) {
```

Comment:
Yeah, the definition of the lambda function for `flatMap` does not throw checked exception.

Sure, I could change it to return `null` on exception instead of empty array. However, the consumer of the output (i.e., constructor of `DatasourceInputSplit`) will detects if `locations` is `null` and assign empty array if so ([here](https://github.com/druid-io/druid/blob/b578adacae978747e15f91acd3b560a40f40a3c5/indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceInputSplit.java#L50)). Should I retain the original returning of `null`? I just feel that we should probably use as little `null` as possible 🤔